### PR TITLE
Merge milestones/M139 to releases/shipped

### PR DESCRIPTION
**Pull request by jrbriggs**
# Release Notes
## Performance Improvements
- GVFS will prefetch commits and trees from a cache server in the background on a schedule, to improve the speed of user-initiated fetches, pulls, and pushes.
- GVFS will pre-calculate status in the background in response to changes in the workdir, to improve the speed of user-initiated statuses.
## General Improvements
- Fixed a bug that would cause GVFS mount process to shut down if it receives out-of-order ProjFS callbacks.

### CODE REVIEW

File: GVFS/GVFS.Installer/Setup.iss [hunk 1]
```diff
@@ -278,6 +283,18 @@ begin
     end;
 end;
 
+procedure WriteGitStatusCacheAvailableFile();
+var
+  TokenFilePath: string;
+begin
+  TokenFilePath := ExpandConstant('{app}\GitStatusCacheAvailable');
```
Comment (sanoursa) [inline GVFS/GVFS.Installer/Setup.iss]: @jamill @jrbriggs  As I mentioned in another comment, this pattern of having the installer write out a file that is specific to a single feature does not seem like a sustainable model. I think all you really want is to write out a file indicating when the upgrade happened, because the question you're trying to answer is: was there a reboot since the last upgrade? That question is not specific to the GitStatusCache.

To be more precise, what the GitStatusCache cares about is: did the last upgrade cross over version X (16 in this case) and has there been a reboot since then?

I think you can do this with an upgrades.dat file (in the format of our `FileBasedDictionary`) that records the time of each upgrade. The GitStatusCache can then query this file and the system reboot time, find the upgrade date/time for the first upgrade that got to 16 or greater, and make sure there has been a reboot since then. There doesn't seem to be a need for a second token file either.

Thoughts on that alternative approach?

File: GVFS/GVFS.Installer/Setup.iss [hunk 2]
```diff
@@ -278,6 +283,18 @@ begin
     end;
 end;
 
+procedure WriteGitStatusCacheAvailableFile();
```
Comment (sanoursa) [inline GVFS/GVFS.Installer/Setup.iss]: In general, I think this pattern is currently too complex, and unfortunately not something we can add functional tests for because of the need to reboot to trigger the behavior. I would definitely like to see more manual testing to ensure that it is working, and it would be nice to add functional tests that at least cover all of the logic. E.g. query the last reboot time, and then (following my upgrades.dat pattern), set up the last upgrade time to be before the actual reboot time vs after the actual reboot time, and confirm that the GitStatusCache responds correctly.

File: GVFS/GVFS.Common/Prefetch/CommitPrefetcher.cs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            // We make a best-effort request to run MIDX and commit-graph writes
+            using (NamedPipeClient pipeClient = new NamedPipeClient(enlistment.NamedPipeName))
+            {
+                if (!pipeClient.Connect())
+                {
+                    tracer.RelatedWarning(
+                        metadata: null,
+                        message: "Failed to connect to GVFS. Skipping post-fetch job request.",
+                        keywords: Keywords.Telemetry);
+                    return true;
```
[the remark this answers is not in the document]
Comment (derrickstolee) [reply GVFS/GVFS.Common/Prefetch/CommitPrefetcher.cs]: The comment above says "best effort" Just because we failed to trigger a background midx/graph computation doesn't mean we should fail the prefetch.

[the remark this answers is not in the document]
Comment (sanoursa) [reply GVFS/GVFS.Common/Prefetch/CommitPrefetcher.cs]: I get that we want it to be best effort, but the code is currently inconsistent. We return true if we fail to connect, but we return false if we fail to get a response. Why do those two code paths behave differently?

Shouldn't this method simply return an honest answer about whether it succeeded or not? The decision of whether to fail the prefetch or not is a business decision that should be driven by `PrefetchVerb`, e.g. by ignoring the return value.

File: GVFS/GVFS.Installer/Setup.iss (hunk 1 above)
Comment (jamill) [reply GVFS/GVFS.Installer/Setup.iss]: I am open to changes - I just want to clarify some things that might not be clear from the naming, as I think the current names are overloaded, and that might be causing confusion.

**Concepts:**
1) When was a version of GVFS that supports the GitStatusCache installed - this is currently recorded by the creation time of `{app}\GitStatusCacheAvailable`.

2) Is the GitStatusCache feature available (has the system been rebooted since a version of GVFS that supports the GitStatusCache was installed). This is `EnableGitStatusCacheToken.dat` that is in the GVFS.Service app data directory.

The installer writes out a file so we can get an accurate time of when a version of GVFS that supports the GitStatusCache was installed (concept 1). That is the purpose of this file - my intent with the naming here was "a file that is created when a version of GVFS that supports GitStatusCache was installed", or "the 1st time a version of GVFS 16 or higher was installed". 

The GVFS.Service will query when the a suitable version of GVFS was installed (by reading the time stamp off the file laid down by the installer), and the last time the system was rebooted. If the system has been rebooted since GVFS was installed, then it will write out a file indicating the GitStatusCache is available (i.e.. concept 2).

The GVFS mount process only checks for the existence of the `EnableGitStatusCacheToken.dat` to make decisions on the availability of the GitStatusCache. We can add functional tests around that GitStatusCache correctly respects the existence of this file with the current approach. Based on previous discussions, the consensus was to keep the logic for comparing the installation time with the reboot time in the GVFS.Service, and having the logic in the mount processes be as simple as possible (i.e. only check for existence of the `EnableGitStatusCacheToken.dat` file. There was other discussion on whether this logic should be platform aware).

**Questions**:
In your suggested approach, is `upgrades.dat` a repository specific file, or is there 1 per GVFS installation? Who would update / maintain this file?

Comment (sanoursa) [reply GVFS/GVFS.Installer/Setup.iss]: Thanks, yea another point of confusion that we have is that we often mix up "product upgrades", meaning updating the bits installed on the box, and "repo upgrades", meaning updating the file formats within a single repo.

In my suggestion, I was referring to product upgrades, and so I would rephrase that as: create a single `ProductUpgrades.dat` file in the service's ProgramData folder. The installer would update this file, and then `GVFS.Mount` would query it to decide if there has been a reboot since we got to version 16.

@jrbriggs  just suggested a simplification of that approach. Instead of maintaining a database of every upgrade, we can just drop a file every time we get a feature that requires a reboot. In this case, the file we drop would be something like `OnDiskVersion16CapableInstallation.dat`, and its create time would indicate when we got from a pre-16 version to a 16-or-greater version. The benefit of this is that the installer doesn't have to do this work on every single upgrade, but on the off chance that we get this sort of requirement again in the future, we have a generic pattern to use.

With both of these new approaches, there is no need for a second token file. The mount process can just query for "when did we get to a version 16 capable installation" and "when did we reboot" and compare those times.

You can still enable functional testing by simply querying the actual reboot time, and then constructing that version-16 file with an appropriate timestamp on it.

File: GVFS/GVFS.Common/Prefetch/CommitPrefetcher.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (derrickstolee) [reply GVFS/GVFS.Common/Prefetch/CommitPrefetcher.cs]: @jrbriggs is this something you can address in #184, since you are already in the prefetcher code?

[the remark this answers is not in the document]
Comment (jrbriggs) [reply GVFS/GVFS.Common/Prefetch/CommitPrefetcher.cs]: @derrickstolee I'm happy to address this, but I'll do it in a separate PR

[the remark this answers is not in the document]
Comment (jrbriggs) [reply GVFS/GVFS.Common/Prefetch/CommitPrefetcher.cs]: Addressing this in #188
